### PR TITLE
[DO-518] How to send payment with Gevamu

### PR DESCRIPTION
**Pull request by StanG4313**

### CODE REVIEW

File: content/4.examples/send-payment.draft.md [hunk 1]
```diff
@@ -12,4 +12,75 @@ classDiagram
...
+    paymentInstruction: PaymentInstruction,
+    gateway: Party
+  ): List<StateAndRef<Payment>> {
+    val flowHandle = serviceHub.startFlow(PaymentFlow(paymentInstruction, gateway, UUID.randomUUID()))
+    return flowHandle.returnValue.get()
+  }
+}
+```
+
+To make a transaction as a network participant, describe Payment in one of formats listed in [`PaymentInstructionFormat`](https://gevamu.github.io/corda-payments-sdk/payments-workflows/com.gevamu.corda.flows/-payment-instruction-format/index.html) enum.
```
Comment (d0rich) [inline content/4.examples/send-payment.draft.md:49]: This sentence can easily continue paragraph before diagram. Let's move it.

**Review by d0rich — CHANGES_REQUESTED** [review]
One small change is required. Also add information about attachments.

File: content/4.examples/send-payment.draft.md (hunk 1 above)
Comment (StanG4313) [reply content/4.examples/send-payment.draft.md]: Did it in new commit

**Review by d0rich — APPROVED** [review]

**Comment by d0rich** [comment]
@StanG4313 in order to make page published in production, you need to remove `.draft` postfix from the name of file.

File: content/4.examples/send-payment.draft.md [hunk 2]
```diff
@@ -12,4 +12,73 @@ classDiagram
...
+  ): List<StateAndRef<Payment>> {
+    val flowHandle = serviceHub.startFlow(PaymentFlow(paymentInstruction, gateway, UUID.randomUUID()))
+    return flowHandle.returnValue.get()
+  }
+}
+```
+
+## Transactions lifecycle
+
+Once payment instruction is received, the Gevamu Payment Solution makes an authentication (using node identity and the Participant ID).
```
Comment (ASitnikov) [inline content/4.examples/send-payment.draft.md:51]:
```suggestion
Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.
```

File: content/4.examples/send-payment.draft.md [hunk 3]
```diff
@@ -12,4 +12,73 @@ classDiagram
...
+If the validation passes, the Payment Gateway returns the ‘Sent to Gateway’ status back to the Participant.
```
Comment (ASitnikov) [inline content/4.examples/send-payment.draft.md:52]:
```suggestion
If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. Status change is communicated back to the Participant's node.
```

File: content/4.examples/send-payment.draft.md [hunk 4]
```diff
@@ -12,4 +12,73 @@ classDiagram
...
+  PCD->>CA: Payment status: Created
+  PCD->>PCD: Passive payment update...
+  CA->>PCD: Query payment
+  PCD->>CA: Payment with relevant status
+
+```
+
+```mermaid
+
+stateDiagram-v2
```
Comment (ASitnikov) [inline content/4.examples/send-payment.draft.md:76]: There is probably the Pending status. I need to check it.

File: content/4.examples/send-payment.draft.md [hunk 5]
```diff
@@ -12,4 +12,73 @@ classDiagram
...
+sequenceDiagram
+  participant CA as MyPaymentService
+  participant PCD as GevamuFacade
+
+  CA->>PCD: Initiate Payment
+  PCD->>PCD: Initiate payment <br/> within PSP
+  PCD->>CA: Payment status: Created
+  PCD->>PCD: Passive payment update...
+  CA->>PCD: Query payment
+  PCD->>CA: Payment with relevant status
```
Comment (ASitnikov) [inline content/4.examples/send-payment.draft.md:70]: I'm not sure it's a part of the flow.

**Review by ASitnikov — CHANGES_REQUESTED** [review]

File: content/4.examples/send-payment.draft.md (hunk 4 above)
Comment (d0rich) [reply content/4.examples/send-payment.draft.md]: @ASitnikov pending status [exists](https://gevamu.github.io/corda-payments-sdk/payments-contracts/com.gevamu.corda.states/-payment/-payment-status/index.html), but as I remember, it exists only inside Gateway and Participant can't see it.

File: content/4.examples/send-payment.draft.md (hunk 5 above)
Comment (d0rich) [reply content/4.examples/send-payment.draft.md]: It is not a part of flow, but it is what user should do in order to get relevant status. @StanG4313 perhaps we need to separate this part of diagram somehow

File: content/4.examples/send-payment.draft.md (hunk 4 above)
Comment (ASitnikov) [reply content/4.examples/send-payment.draft.md]: It's visible to the payer (participant)

File: content/4.examples/send-payment.draft.md (hunk 5 above)
Comment (ASitnikov) [reply content/4.examples/send-payment.draft.md]: It's possible to subscribe to vault updates with `trackBy()` https://docs.r3.com/en/platform/corda/4.9/enterprise/cordapps/api-vault-query.html

I think there was also a possibility to start a flow when corda state changes, but I cannot quickly find it.

Comment (StanG4313) [reply content/4.examples/send-payment.draft.md]: Fixed in new commit

File: content/4.examples/send-payment.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  PCD->>PCD: Passive payment update...
+  Note over CA,PCD: After state change
+  CA->>PCD: Query payment
+  PCD->>CA: Payment with relevant status
+
+```
+
+```mermaid
+
+stateDiagram-v2
```
Comment (d0rich) [inline content/4.examples/send-payment.md:77]: Add full name for Sent status.
```suggestion
stateDiagram-v2
    Sent: Sent to gateway
```

**Review by d0rich — CHANGES_REQUESTED** [review]
One small change required

File: content/4.examples/send-payment.md (hunk 6 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md:77]: Is it an output are you going to have, @d0rich ? 
image

Comment (d0rich) [reply content/4.examples/send-payment.md:77]: Nope, this output:

image

Full code:

```mermaid
stateDiagram-v2
    Sent: Sent to gateway
    [*] --> Created
    Created --> Sent
    Sent --> Pending
    Pending --> Accepted
    Pending --> Rejected
    Accepted --> [*]
    Rejected --> [*]
```

File: content/4.examples/send-payment.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
+# Send Payment
+
+This example shows how to send a payment from one account to another via Gevamu Payments Solution.
+
+```mermaid
+classDiagram
+  class MyPaymentService {
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:7]: Do we need `MyPaymentService` in this example?

File: content/4.examples/send-payment.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+
+class GevamuFacade(private val serviceHub: AppServiceHub): SingletonSerializeAsToken() {
+
+  fun sendPayment(
+    paymentInstruction: PaymentInstruction,
+    gateway: Party
+  ): List<StateAndRef<Payment>> {
+    val flowHandle = serviceHub.startFlow(PaymentFlow(paymentInstruction, gateway, UUID.randomUUID()))
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:43]: If I'm not mistaken, the last parameter (UUID) is optional. If it is I'd remove it in this example.

File: content/4.examples/send-payment.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        gevamuFacade.sendPayment(paymentInstruction, gateway)
+    }
+}
+
+class GevamuFacade(private val serviceHub: AppServiceHub): SingletonSerializeAsToken() {
+
+  fun sendPayment(
+    paymentInstruction: PaymentInstruction,
+    gateway: Party
+  ): List<StateAndRef<Payment>> {
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:42]: It's worth explaining why `List` is returned from the `PaymentFlow`

File: content/4.examples/send-payment.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    val flowHandle = serviceHub.startFlow(PaymentFlow(paymentInstruction, gateway, UUID.randomUUID()))
+    return flowHandle.returnValue.get()
+  }
+}
+```
+
+## Transactions lifecycle
+
+Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.
+If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. Status change is communicated back to the Participant's node.
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:52]: It relates to my previous comment. If payment initiation request has multiple payment instructions, multiple payment states are created and are independently updated.

File: content/4.examples/send-payment.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+Transaction can be accepted or rejected by the Payment Service Provider.
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:53]: If payment cannot be immediately completed, "Pending" status is returned. The payment state will be updated with the final status when the Gevamu gateway receives new status from the PSP.

File: content/4.examples/send-payment.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+To make a transaction as a network participant, describe Payment in one of formats listed in [`PaymentInstructionFormat`](https://gevamu.github.io/corda-payments-sdk/payments-workflows/com.gevamu.corda.flows/-payment-instruction-format/index.html) enum.
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:55]: For me this phrase implies that it's possible to make a transaction not as a network participant :-)

File: content/4.examples/send-payment.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+stateDiagram-v2
+    Sent: Sent to gateway
+    [*] --> Created
+    Created --> Sent
+    Sent --> Pending
+    Pending --> Accepted
+    Pending --> Rejected
+    Accepted --> [*]
+    Rejected --> [*]
```
Comment (ASitnikov) [inline content/4.examples/send-payment.md:85]:
```suggestion
    [*] --> Created
    Created --> Sent
    Sent --> Pending
    Sent  --> Accepted
    Sent  --> Rejected
    Pending --> Accepted
    Pending --> Rejected
    Accepted --> [*]
    Rejected --> [*]
```

**Review by ASitnikov — CHANGES_REQUESTED** [review]

File: content/4.examples/send-payment.md (hunk 9 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md:37]: Explanation added

File: content/4.examples/send-payment.md (hunk 11 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md:48]: Explanation added

File: content/4.examples/send-payment.md (hunk 10 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md]: Explanation added

File: content/4.examples/send-payment.md (hunk 8 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md:38]: That's true, will remove it in the next commit

File: content/4.examples/send-payment.md (hunk 12 above)
Comment (StanG4313) [reply content/4.examples/send-payment.md:53]: Fixed in next commit

**Review by d0rich — CHANGES_REQUESTED** [review]

File: content/4.examples/send-payment.md [hunk 14]
```diff
@@ -49,10 +44,15 @@ class GevamuFacade(private val serviceHub: AppServiceHub): SingletonSerializeAsT
 ## Transactions lifecycle
 
 Once payment instruction is received, the Gevamu Payment Solution validates and authorizes (by checking node identity and the Participant ID) the payment.
-If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. Status change is communicated back to the Participant's node.
+If the validation passes, the Payment Gateway sets the payment status to ‘Sent to Gateway’. 
+Status change is communicated back to the Participant's node.
+Some payment initiation requests has multiple payment instructions.
+In this case multiple payment states are created and independently updated.
```
[the remark this answers is not in the document]
Comment (lingvolab) [reply content/4.examples/send-payment.md:50]: If a payment initiation request contains multiple payment instructions, multiple payment states will be created and independently updated. 
(singular/plural nouns should be grammatically aligned with the verb)

File: content/4.examples/send-payment.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+```mermaid
+classDiagram
+  class GevamuFacade {
+    + sendPayment()
+  }
+```
+
+## Program implementation
+
+Sending payment is quite simple operation. It includes only call of `PaymentFlow`.
```
Comment (lingvolab) [inline content/4.examples/send-payment.md:14]: subjective statement - please rephrase

File: content/4.examples/send-payment.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (StanG4313) [reply content/4.examples/send-payment.md:50]:
```suggestion
If payment initiation requests contain multiple payment instructions, multiple payment states will be created and independently updated.
```
Now better?

File: content/4.examples/send-payment.md (hunk 15 above)
Comment (lingvolab) [reply content/4.examples/send-payment.md:14]:
```suggestion
Sending a payment is a one-step operation, consisting of a single `PaymentFlow` call. 
```

File: content/4.examples/send-payment.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (lingvolab) [reply content/4.examples/send-payment.md:50]:
```suggestion
If a payment initiation request contains multiple payment instructions, multiple payment states will be created and independently updated.
```

**Review by lingvolab — APPROVED** [review]

**Review by d0rich — APPROVED** [review]

**Review by ASitnikov — APPROVED** [review]
Looks good to me.

**Pull request merged**